Sources/SwiftParser/StringLiterals.swift

```diff
@@ -0,0 +1,126 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+@_spi(RawSyntax) import SwiftSyntax
+
+extension Parser {
+  /// Consumes a raw string delimiter that has the same number of `#` as `openDelimiter`.
+  private mutating func parseStringDelimiter(openDelimiter: RawTokenSyntax?) -> (unexpectedBeforeCheckedDelimiter: RawUnexpectedNodesSyntax?, checkedDelimiter: RawTokenSyntax?) {
+    // Check for leadingTriviaText == "" so we don't consume the leading raw
+    // string delimiter of an upcoming string literal, e.g. in
+    // ```
+    // "normal literal"
+    // #"raw literal"#
+    // ```
+    let delimiter: RawTokenSyntax?
+    if self.at(.rawStringDelimiter) && self.currentToken.leadingTriviaText == "" {
+      delimiter = self.consumeAnyToken()
+    } else {
+      delimiter = nil
+    }
+
+    switch (openDelimiter, delimiter) {
+    case (nil, nil):
+      return (nil, nil)
+    case (let open?, nil):
+      return (nil, missingToken(.rawStringDelimiter, text: open.tokenText))
+    case (nil, .some):
+      return (RawUnexpectedNodesSyntax([delimiter], arena: self.arena), nil)
+    case (let open?, let close?):
+      if open.tokenText == close.tokenText {
+        return (nil, close)
+      } else {
+        return (RawUnexpectedNodesSyntax([delimiter], arena: self.arena), missingToken(.rawStringDelimiter, text: open.tokenText))
+      }
+    }
+  }
+
+  /// Parse a string literal expression.
+  @_spi(RawSyntax)
+  public mutating func parseStringLiteral() -> RawStringLiteralExprSyntax {
+    /// Parse opening raw string delimiter if exist.
+    let openDelimiter = self.consume(if: .rawStringDelimiter)
+
+    /// Parse open quote.
+    let (unexpectedBeforeOpenQuote, openQuote) = self.expectAny([.stringQuote, .multilineStringQuote, .singleQuote], default: .stringQuote)
+
+    /// Parse segments.
+    var segments: [RawStringLiteralSegmentsSyntax.Element] = []
+    var loopProgress = LoopProgressCondition()
+    while loopProgress.evaluate(self.currentToken) {
+      if let stringSegment = self.consume(if: .stringSegment) {
+        segments.append(.stringSegment(RawStringSegmentSyntax(content: stringSegment, arena: self.arena)))
+      } else if let backslash = self.consume(if: .backslash) {
+        let (unexpectedBeforeDelimiter, delimiter) = self.parseStringDelimiter(openDelimiter: openDelimiter)
+        let (unexpectedBeforeLeftParen, leftParen) = self.expect(.leftParen)
+        let expressions = RawTupleExprElementListSyntax(elements: self.parseArgumentListElements(pattern: .none), arena: self.arena)
```

Comment:
How newlines in the string interpolation are handled? e.g.
```
let a = "test \(label:
foo()
```
This should be missing expression after `label:`, missing `)`, and missing `"`. But I think `foo()` is parsed as the expression, no?

Reply:
We are leaving the `inStringInterpolationState` state when hitting a newline because of

https://github.com/apple/swift-syntax/pull/1250/files#diff-253b01bc981faa185173c1697784a41bf1b586e91ec7b4de806dfd2646db52c3R135

I added test additional test cases for this